src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/CollectBaseDistributionByCycleSpark.java

```diff
@@ -0,0 +1,240 @@
+                    return 2;
+                case 'T':
+                case 't':
+                    return 3;
+            }
+            return 4;
+        }
+
+        HistogramGenerator addRead(final GATKRead rec) {
+            final byte[] bases = rec.getBases();
+            if (bases == null) {
+                return this;
+            }
+            final int length = bases.length;
+            final boolean rc = rec.isReverseStrand();
+            ensureArraysBigEnough(length + 1);
+            if (rec.isPaired() && rec.isSecondOfPair()) {
+                seenSecondEnd = true;
+                for (int i = 0; i < length; i++) {
+                    final int cycle = rc ? length - i : i + 1;
+                    secondReadTotalsByCycle[baseToInt(bases[i])][cycle] += 1;
+                    secondReadCountsByCycle[cycle] += 1;
+                }
+            } else {
+                for (int i = 0; i < length; i++) {
+                    final int cycle = rc ? length - i : i + 1;
+                    firstReadTotalsByCycle[baseToInt(bases[i])][cycle] += 1;
+                    firstReadCountsByCycle[cycle] += 1;
+                }
+            }
+            return this;
+        }
+
+        private void ensureArraysBigEnough(final int length) {
+            if (length > maxLengthSoFar) {
+                for (int i = 0; i < NBASES; i++) {
+                    firstReadTotalsByCycle[i] = Arrays.copyOf(firstReadTotalsByCycle[i], length);
+                    secondReadTotalsByCycle[i] = Arrays.copyOf(secondReadTotalsByCycle[i], length);
+                }
+                firstReadCountsByCycle = Arrays.copyOf(firstReadCountsByCycle, length);
+                secondReadCountsByCycle = Arrays.copyOf(secondReadCountsByCycle, length);
+                maxLengthSoFar = length;
+            }
+        }
+
+        boolean isEmpty() {
+            return maxLengthSoFar == 0;
+        }
+
+        public void addToMetricsFile(final MetricsFile<BaseDistributionByCycleMetrics, ?> metrics) {
+            int firstReadLength = 0;
+            for (int i = 0; i < maxLengthSoFar; i++) {
+                if (0 != firstReadCountsByCycle[i]) {
+                    final BaseDistributionByCycleMetrics metric = new BaseDistributionByCycleMetrics();
+                    metric.READ_END = 1;
+                    metric.CYCLE = i;
+                    metric.PCT_A = (100.0 * firstReadTotalsByCycle[0][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_C = (100.0 * firstReadTotalsByCycle[1][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_G = (100.0 * firstReadTotalsByCycle[2][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_T = (100.0 * firstReadTotalsByCycle[3][i] / firstReadCountsByCycle[i]);
+                    metric.PCT_N = (100.0 * firstReadTotalsByCycle[4][i] / firstReadCountsByCycle[i]);
+                    metrics.addMetric(metric);
+                    firstReadLength = i;
```

Comment:
This isn't really firstReadLength (it's 1 more than that), it's really firstReadCycles.